analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabaseCPlugin.java

```diff
@@ -0,0 +1,234 @@
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import org.pf4j.Extension;
+import org.pf4j.Plugin;
+import org.pf4j.PluginWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.sql.BatchUpdateException;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class MetadataDatabaseCPlugin extends Plugin {
+    public MetadataDatabaseCPlugin(PluginWrapper wrapper) {
+        super(wrapper);
+    }
+
+    @Extension
+    public static class MetadataDBCExtension extends MetadataDBExtension {
+        private static DSLContext dslContext;
+
+        @Override
+        public void setDBConnection(DSLContext dslContext) {
+            MetadataDBCExtension.dslContext = dslContext;
+        }
+
+        @Override
+        public DSLContext getDBConnection() {
+            return MetadataDBCExtension.dslContext;
+        }
+
+        /**
+         * Saves a callgraph of new format to the database to appropriate tables.
+         * We override this method because we want to save the architecture
+         * of the package.
+         *
+         * @param callGraph   Call graph to save to the database.
+         * @param metadataDao Data Access Object to insert records in the database
+         * @return Package ID saved in the database
+         */
+        protected long saveToDatabase(ExtendedRevisionCallGraph callGraph, MetadataDao metadataDao) {
+            ExtendedRevisionCCallGraph CCallGraph = (ExtendedRevisionCCallGraph) callGraph;
+            // Insert package record
+            final long packageId = metadataDao.insertPackage(CCallGraph.product, CCallGraph.forge);
+
+            // Insert package version record
+            final long packageVersionId = metadataDao.insertPackageVersion(packageId,
+                    CCallGraph.getCgGenerator(), CCallGraph.version, CCallGraph.architecture,
+                    getProperTimestamp(CCallGraph.timestamp), new JSONObject());
+
+            var callables = instertDataExtractCallables(callGraph, metadataDao, packageVersionId);
+            final var numInternal = callables.size();
+
+            var callablesIds = new LongArrayList(callables.size());
+            // Save all callables in the database
+            callablesIds.addAll(metadataDao.insertCallablesSeparately(callables, numInternal));
+
+            // Build a map from callable Local ID to Global ID
+            var lidToGidMap = new Long2LongOpenHashMap();
+            for (int i = 0; i < callables.size(); i++) {
+                lidToGidMap.put(callables.get(i).getId().longValue(), callablesIds.getLong(i));
+            }
+
+            // Insert all the edges
+            var edges = insertEdges(callGraph.getGraph(), lidToGidMap, metadataDao);
+
```

Comment:
I think you forgot to add code to remove duplicate nodes, see `MetadataDBExtension` lines 252-266

Reply:
https://github.com/fasten-project/fasten/pull/91/commits/1b7faf155d3d136ef602687bf6537c5981a55cdf